packages/cursorless-vscode/src/CommandHistory.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  ActionDescriptor,
+  CommandComplete,
+  CommandHistoryItem,
+  Disposable,
+  FileSystem,
+  ReadOnlyHatMap,
+} from "@cursorless/common";
+import type {
+  CommandRunner,
+  CommandRunnerDecorator,
+} from "@cursorless/cursorless-engine";
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import * as vscode from "vscode";
+
+const filePrefix = "cursorlessCommandHistory";
+const settingSection = "cursorless";
+const settingName = "commandHistory";
+const settingFullName = `${settingSection}.${settingName}`;
+
+export class CommandHistory implements CommandRunnerDecorator {
+  private readonly dirPath: string;
+  private readonly cursorlessVersion: string;
+  private disposable: Disposable;
+  private isActive: boolean = false;
+
+  constructor(
+    extensionContext: vscode.ExtensionContext,
+    fileSystem: FileSystem,
+  ) {
+    this.cursorlessVersion = extensionContext.extension.packageJSON.version;
+    this.dirPath = fileSystem.cursorlessCommandHistoryDirPath;
+
+    // Read initial setting value. The watcher below will take care of changes.
+    this.evaluateSetting();
+
+    this.disposable = vscode.workspace.onDidChangeConfiguration((event) => {
+      if (event.affectsConfiguration(settingFullName)) {
+        this.evaluateSetting();
+      }
+    });
+  }
+
+  wrapCommandRunner(
+    readableHatMap: ReadOnlyHatMap,
+    runner: CommandRunner,
+  ): CommandRunner {
+    if (!this.isActive) {
+      return runner;
+    }
+
+    return {
+      run: async (commandComplete: CommandComplete) => {
+        await this.appendToLog(commandComplete);
+
+        return await runner.run(commandComplete);
+      },
+    };
+  }
+
+  private async appendToLog(command: CommandComplete): Promise<void> {
+    const date = new Date();
+    const fileName = `${filePrefix}_${getMonthDate(date)}.jsonl`;
+    const file = path.join(this.dirPath, fileName);
+
+    const historyItem: CommandHistoryItem = {
+      date: getDayDate(date),
+      cursorlessVersion: this.cursorlessVersion,
+      command: sanitizeCommand(command),
```

Comment:
Probably not worth changing, but `sanitizeCommand` and `sanitizeAction` would prob be cleaner using immer, like we do with canonicalization. You'd basically just change those functions to mutate the commands directly, calling them eg `sanitizeCommandInPlace`, then just use `produce` here like we do in https://github.com/cursorless-dev/cursorless/blob/40a6fee4c4eec9b37cfb1ee7b4bf5ff78df59ea2/packages/cursorless-engine/src/core/commandVersionUpgrades/canonicalizeAndValidateCommand.ts#L40-L46

Reply:
I don't know. Personally I find that pattern quite hard to read.

Reply:
I find deeply nested spreads hard to read. What don't you like about the immer pattern?

Just to be clear, let's not block this PR on this discussion, but I'm interested to have it anyway

Reply:
That is just so the code should be as compact as possible. You can definitely make it more verbose if you prefer that pattern. 

I think that immer adds a layer of abstraction that is quite esoteric. Instead of just returning or assigning a value all of the sudden you have a callback with an argument and no return value. Instead of constructing a new object without a field you should now do `delete draft.spokenForm`. In my mind immer solves a problem that we don't really have here and goes against the kiss principle.

Reply:
Let's not block this PR on this discussion. I did a quick draft of what this would look like with immer in https://github.com/cursorless-dev/cursorless/pull/2124; let's discuss that one at our leisure. Not urgent at all